apps/src/panels/PanelsView.tsx

```diff
@@ -0,0 +1,163 @@
+
+  const levelData = useSelector(
+    ({lab}: {lab: LabState}) => lab.levelProperties?.levelData
+  );
+  const currentAppName = useSelector(
+    ({lab}: {lab: LabState}) => lab.levelProperties?.appName
+  );
+
+  const [levelPanels, setLevelPanels] = useState<PanelsLevelData | null>(null);
+  const [currentPanel, setCurrentPanel] = useState(0);
+
+  useEffect(() => {
+    if (currentAppName === appName && levelData) {
+      setLevelPanels(levelData as PanelsLevelData);
+    }
+  }, [currentAppName, levelData]);
+
+  const handleButtonClick = useCallback(() => {
+    if (levelPanels?.panels) {
+      if (currentPanel < levelPanels.panels.length - 1) {
+        setCurrentPanel(currentPanel + 1);
+      } else {
+        dispatch(sendSuccessReport(appName));
+        dispatch(navigateToNextLevel());
+      }
+    }
+  }, [levelPanels, currentPanel, dispatch]);
+
+  const handleBubbleClick = (index: number) => {
+    setCurrentPanel(index);
+  };
+
+  // Leave a margin to the left and the right of the panels, to the edges
+  // of the screen.
+  const horizontalMargin = 40;
+
+  // Leave a vertical margin above and below the panels, to the edges of the
+  // screen.
+  const verticalMargin = 50;
+
+  // We need room below the panels content for the children passed in.  This area
+  // can contain things like a Continue button.
+  const childrenAreaHeight = 70;
+
+  // The aspect ratio of the panels.
+  const contentAspectRatio = 16 / 9;
+
+  let [targetWidth, targetHeight] = useWindowSize();
+  targetWidth -= horizontalMargin * 2;
+  targetHeight -= verticalMargin * 2 + childrenAreaHeight;
+
+  let width, height;
+  if (targetWidth / targetHeight > contentAspectRatio) {
+    height = targetHeight;
+    width = contentAspectRatio * height;
+  } else {
+    width = targetWidth;
+    height = width / contentAspectRatio;
+  }
+
+  if (!levelPanels?.panels) {
+    return <div />;
+  }
+
+  return (
+    <div
+      id="panels-container"
+      className={styles.panelsContainer}
+      key={currentPanel}
+    >
+      <div className={styles.panel} style={{width, height}}>
+        <div
+          className={styles.image}
+          style={{
+            backgroundImage: `url("${levelPanels.panels[currentPanel].imageUrl}")`,
+          }}
+        />
+        <EnhancedSafeMarkdown
+          markdown={levelPanels.panels[currentPanel]?.text}
+          className={styles.markdownText}
+        />
+      </div>
+      <div
+        className={styles.childrenArea}
+        style={{width: width, height: childrenAreaHeight}}
+      >
+        <button
+          id="panels-button"
+          type="button"
+          onClick={handleButtonClick}
+          className={styles.button}
+        >
+          {currentPanel < levelPanels.panels.length - 1
+            ? commonI18n.next()
+            : commonI18n.continue()}
+        </button>
+        <div id="panels-bubbles">
+          {Array.from(Array(levelPanels.panels.length).keys()).map(index => {
```

Comment:
I've seen this kind of index generator in a couple of places (`Array.from(Array(length).keys()`) - worth pulling out into a util later?

Reply:
Oh, we have [getRangeArray.ts ](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/dance/ai/utils/getRangeArray.ts)now.  Maybe we could move it to `apps/src/util/`?